src/operator/communication/shard.rs

```diff
@@ -230,6 +230,47 @@ where
     }
 }
 
+impl<P, T> Stream<Circuit<P>, T>
+where
+    P: Clone + 'static,
+    T: 'static,
+{
+    /// Marks the data within the current stream as sharded, meaning that all
+    /// further calls to `.shard()` will have no effect.
+    ///
+    /// This must only be used on streams of values that are properly sharded
+    /// across workers, otherwise this will cause the dataflow to yield
+    /// incorrect results
+    pub fn mark_sharded(&self) -> Self {
+        self.circuit().cache_insert(
+            ShardId::new((
+                self.origin_node_id().clone(),
+                sharding_policy(self.circuit()),
+            )),
+            self.clone(),
+        );
+
+        self.clone()
+    }
+
+    /// Returns `true` if the current stream is properly sharded across workers
+    pub(crate) fn is_sharded(&self) -> bool {
+        Runtime::runtime()
+            .map(|runtime| {
+                if runtime.num_workers() == 1 {
+                    true
+                } else {
+                    self.circuit()
+                        .cache_contains(&ShardId::<Circuit<P>, T>::new((
```

Comment:
This can return `true` is `self` is not sharded, but there is another stream created by `self.shard()` that carries the sharded version of `self`.

Reply:
I think I fixed this with `DirectShardId` but I'm not 100% sure

Reply:
This seems too heavy-handed.  We shouldn't need another cache id for this. How about this solution:
- `shard_generic` records `Cache[ShardId(sharded)] := sharded` in addition to `Cached[ShardId(self)] := sharded`.
- `mark_sharded` records `Cache[ShardId(self)] := self`.
- All operators that preserve "shardedness" of inputs:
  - if `let Some(sharded) = Cache[ShardId(stream)]` then use `sharded` instead of `self` as input and mark output as sharded.
- remove `is_sharded` method, as it answers the wrong question: we don't normally care if the stream is sharded, but rather whether there exists a sharded version of the stream.

Reply:
I considered that but I'm not sure I like it to be honest, it seems like just a little bit too much magic to shard operators that weren't previously sharded and don't technically need sharding

Reply:
We certainly don't want to do that, but that's not what I'm suggesting.  The operator should simply check if there already _exists_ a cached sharded version of the stream and use that if possible.  Otherwise it should just work with non-sharded inputs and produce non-sharded outputs.